src/hotspot/share/utilities/bitMap.cpp

```diff
@@ -99,85 +64,35 @@ BitMap::bm_word_t* BitMap::reallocate(const Allocator& allocator, bm_word_t* old
   }
 
   if (old_map != NULL) {
-    allocator.free(old_map, old_size_in_words);
+    derived->free(old_map, old_size_in_words);
   }
 
   return map;
 }
 
-template <class Allocator>
-bm_word_t* BitMap::allocate(const Allocator& allocator, idx_t size_in_bits, bool clear) {
+ArenaBitMap::ArenaBitMap(Arena* arena, idx_t size_in_bits, bool clear) : GrowableBitMap<ArenaBitMap>(), _arena(arena) {
   // Reuse reallocate to ensure that the new memory is cleared.
-  return reallocate(allocator, NULL, 0, size_in_bits, clear);
-}
-
-template <class Allocator>
-void BitMap::free(const Allocator& allocator, bm_word_t* map, idx_t  size_in_bits) {
-  bm_word_t* ret = reallocate(allocator, map, size_in_bits, 0);
-  assert(ret == NULL, "Reallocate shouldn't have allocated");
-}
-
-template <class Allocator>
-void BitMap::resize(const Allocator& allocator, idx_t new_size_in_bits, bool clear) {
-  bm_word_t* new_map = reallocate(allocator, map(), size(), new_size_in_bits, clear);
-
-  update(new_map, new_size_in_bits);
-}
-
-template <class Allocator>
-void BitMap::initialize(const Allocator& allocator, idx_t size_in_bits, bool clear) {
-  assert(map() == NULL, "precondition");
-  assert(size() == 0,   "precondition");
-
-  resize(allocator, size_in_bits, clear);
+  bm_word_t* map = reallocate(nullptr, 0, size_in_bits, clear);
+  update(map, size_in_bits);
 }
 
-template <class Allocator>
-void BitMap::reinitialize(const Allocator& allocator, idx_t new_size_in_bits, bool clear) {
-  // Remove previous bits - no need to clear
-  resize(allocator, 0, false /* clear */);
-
-  initialize(allocator, new_size_in_bits, clear);
+bm_word_t* ArenaBitMap::allocate(idx_t size_in_words) const {
+  return (bm_word_t*)_arena->Amalloc(size_in_words * BytesPerWord);
 }
 
 ResourceBitMap::ResourceBitMap(idx_t size_in_bits, bool clear)
-    : BitMap(allocate(ResourceBitMapAllocator(), size_in_bits, clear), size_in_bits) {
-}
-
-void ResourceBitMap::resize(idx_t new_size_in_bits) {
-  BitMap::resize(ResourceBitMapAllocator(), new_size_in_bits, true /* clear */);
-}
-
-void ResourceBitMap::initialize(idx_t size_in_bits) {
-  BitMap::initialize(ResourceBitMapAllocator(), size_in_bits, true /* clear */);
-}
-
-void ResourceBitMap::reinitialize(idx_t size_in_bits) {
-  BitMap::reinitialize(ResourceBitMapAllocator(), size_in_bits, true /* clear */);
-}
-
-ArenaBitMap::ArenaBitMap(Arena* arena, idx_t size_in_bits)
-    : BitMap(allocate(ArenaBitMapAllocator(arena), size_in_bits), size_in_bits) {
+  : GrowableBitMap<ResourceBitMap>(size_in_bits, clear) {
 }
 
 CHeapBitMap::CHeapBitMap(idx_t size_in_bits, MEMFLAGS flags, bool clear)
-    : BitMap(allocate(CHeapBitMapAllocator(flags), size_in_bits, clear), size_in_bits), _flags(flags) {
+  : GrowableBitMap<CHeapBitMap>(), _flags(flags) {
+  // Reuse reallocate to ensure that the new memory is cleared.
+  bm_word_t* map = reallocate(nullptr, 0, size_in_bits, clear);
+  update(map, size_in_bits);
```

Comment:
Both CHeapBitMap and AreanBitMap perform this "initialization". I propose that this instead call the GrowableBitMap::initialize function.

Reply:
I agree! 'initialize()' is clearer to reflect the purpose of code.